### ISSUE TITLE
Running too many VMs hard reboots entire system

### ISSUE THREAD

**Issue by DarkLog1x** [QubesOS/qubes-issues#5082]
**Qubes OS version**
Qubes OS R4.0

**Affected component(s) or functionality**
Dom0

**Brief summary**
Once around 9-12 VMs are started if another VM is started it will cause the entire system to hard reboot. 

**To Reproduce**
1. Start new VMs until the system crashes 

**Expected behavior**
System not hard rebooting.

**Actual behavior**
Once several VMs are running and a new VM is started, the VM will seem to start and then hard crash. The system will start to boot up from bios/uefi.

**Additional context**
This seems to be a memory bug in Dom0. It is closely related to #4891 however I do not know Xen and Qubes memory management therefor cannot pinpoint in more detail. 

**Solutions you've tried**
In #4891 changing the line `` DOM0_MAXMEM=`/usr/sbin/xl list 0 | awk '{ getline; print $3 }'` `` in  file `/usr/lib/qubes/startup-misc.sh` seemed to fix the issue until the last ~2 days when it came back. It seems like the fix for that issue was pushed into the stable branch around the time that this issue came back. 

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**
Issue #4891

**Comment by andrewdavidwong** [comment]
Possibly related to or duplicate of: #3079

**Comment by unman** [comment]
I have an old x230 with 12GB RAM - it's fine. Currently running 14 qubes, and can spin up others as needed.
If I make a concerted effort to overrun memory by just starting qubes one after another, the worst that happens is that some of those qubes never start. This is expected.
This is the most restricted Qubes machine I have - running Qubes stable, mainly Debian templates, some but not all with maxmem 300 - I've never had a hard reboot on it.

@DarkLog1x It might be helpful if you gave details of your hardware, qube settings, and what Qubes software versions you are running.

**Comment by DarkLog1x** [comment]
Hey I have dug more into the issue and seem to have been able to solve it after digging through log output. The tl:dr is that it seems like the nouveau driver fails to change power state and crashes the machine. 

My initial thought that the number of VMs caused the crash was wrong as when I started to test more it seems like crashes occurred with as little as 4 VMs. I therefore dont think the issue is related to #3079 .

For some more clarification I am on a ThinkPad w540 with an Intel CPU/iGPU and a Nvidia chip (Quadro K1100M). I therefore had the nouveau drivers running as they are the ones that were installed. After looking at the last boot logs via `journalctl -b -1` I was able to see that the nouveau driver tries to change power state and then seems to get stuck and crashes the system:

```
Jun 09 15:31:01 dom0 qmemman.daemon.algo[2408]: balance_when_enough_memory(xen_free_memory=10028544086, total_mem_pref=2027570380.8, total_available_memory=12977516425.2)
Jun 09 15:31:01 dom0 qmemman.daemon.algo[2408]: left_memory=7578916237 acceptors_count=1
Jun 09 15:31:01 dom0 qmemman.systemstate[2408]: stat: dom '7' act=629146624 pref=414407884.8 last_target=629146624
Jun 09 15:31:01 dom0 qmemman.systemstate[2408]: stat: dom '0' act=4294967296 pref=1423370240.0 last_target=4294967296
Jun 09 15:31:01 dom0 qmemman.systemstate[2408]: stat: dom '5' act=52428800 pref=189792256 last_target=52428800
Jun 09 15:31:01 dom0 qmemman.systemstate[2408]: stat: xenfree=10080972886 memset_reqs=[('0', 4294967296), ('7', 6291456000), ('5', 52428800)]
Jun 09 15:31:01 dom0 qmemman.systemstate[2408]: mem-set domain 0 to 4294967296
Jun 09 15:31:01 dom0 qmemman.systemstate[2408]: mem-set domain 7 to 6291456000
Jun 09 15:31:01 dom0 qmemman.systemstate[2408]: mem-set domain 5 to 52428800
Jun 09 15:31:02 dom0 kernel: video LNXVIDEO:01: Cannot transition to power state D0 for parent in D3cold
Jun 09 15:31:02 dom0 kernel: nouveau 0000:01:00.0: Refused to change power state, currently in D3
Jun 09 15:31:02 dom0 kernel: video LNXVIDEO:01: Cannot transition to power state D0 for parent in D3cold
Jun 09 15:31:02 dom0 kernel: nouveau 0000:01:00.0: Refused to change power state, currently in D3
Jun 09 15:31:02 dom0 kernel: video LNXVIDEO:01: Cannot transition to power state D0 for parent in D3cold
Jun 09 15:31:02 dom0 kernel: nouveau 0000:01:00.0: Refused to change power state, currently in D3
Jun 09 15:31:02 dom0 kernel: nouveau 0000:01:00.0: tmr: stalled at ffffffffffffffff
Jun 09 15:31:02 dom0 kernel: ------------[ cut here ]------------
Jun 09 15:31:02 dom0 kernel: nouveau 0000:01:00.0: timeout
Jun 09 15:31:02 dom0 kernel: WARNING: CPU: 1 PID: 4129 at /home/user/rpmbuild/BUILD/kernel-latest-5.1.2/linux-5.1.2/drivers/gpu/drm/nouveau/nvkm/subdev/pmu/base.c:86 nvkm_pmu_reset+0x1a6/0x1
Jun 09 15:31:02 dom0 kernel: Modules linked in: loop ebtable_filter ebtables ip6table_filter ip6_tables iptable_filter rmi_smbus rmi_core iTCO_wdt iTCO_vendor_support mei_wdt mei_hdcp intel_
Jun 09 15:31:02 dom0 kernel: CPU: 1 PID: 4129 Comm: Xorg Tainted: G     U            5.1.2-1.pvops.qubes.x86_64 #1
Jun 09 15:31:02 dom0 kernel: Hardware name: LENOVO 20BGCTO1WW/20BGCTO1WW, BIOS GNET83WW (2.31 ) 05/03/2017
Jun 09 15:31:02 dom0 kernel: RIP: e030:nvkm_pmu_reset+0x1a6/0x1c0 [nouveau]
Jun 09 15:31:02 dom0 kernel: Code: 04 24 48 8b 40 10 48 8b 78 10 4c 8b 67 50 4d 85 e4 74 24 e8 dc f6 1c c1 4c 89 e2 48 89 c6 48 c7 c7 a3 75 5f c0 e8 5a 67 c4 c0 <0f> 0b e9 4b ff ff ff 48 8b 
Jun 09 15:31:02 dom0 kernel: RSP: e02b:ffffc90008effb70 EFLAGS: 00010286
Jun 09 15:31:02 dom0 kernel: RAX: 0000000000000000 RBX: ffff8881a063f400 RCX: 0000000000000006
Jun 09 15:31:02 dom0 kernel: RDX: 0000000000000007 RSI: 0000000000000001 RDI: ffff8881aae568c0
Jun 09 15:31:02 dom0 kernel: RBP: ffff88819e946480 R08: 0000000000000001 R09: 0000000000000467
Jun 09 15:31:02 dom0 kernel: R10: 0000000000000000 R11: 0000000000000000 R12: ffff8881a5f95d10
Jun 09 15:31:02 dom0 kernel: R13: 00000a96d71ee73a R14: 00000a96d71ee4ec R15: ffffc90008effdc8
Jun 09 15:31:02 dom0 kernel: FS:  0000779a52843a40(0000) GS:ffff8881aae40000(0000) knlGS:0000000000000000
Jun 09 15:31:02 dom0 kernel: CS:  e030 DS: 0000 ES: 0000 CR0: 0000000080050033
Jun 09 15:31:02 dom0 kernel: CR2: 00005d5d904e8378 CR3: 0000000186190000 CR4: 0000000000042660
Jun 09 15:31:02 dom0 kernel: Call Trace:
Jun 09 15:31:02 dom0 kernel:  nvkm_subdev_preinit+0x30/0xe0 [nouveau]
Jun 09 15:31:02 dom0 kernel:  ? nvkm_device_pci_preinit+0x39/0x50 [nouveau]
Jun 09 15:31:02 dom0 kernel:  nvkm_device_init+0x61/0x280 [nouveau]
Jun 09 15:31:02 dom0 kernel:  nvkm_udevice_init+0x41/0x60 [nouveau]
Jun 09 15:31:02 dom0 kernel:  nvkm_object_init+0x3e/0x190 [nouveau]
Jun 09 15:31:02 dom0 kernel:  nvkm_object_init+0xb1/0x190 [nouveau]
Jun 09 15:31:02 dom0 kernel:  nvkm_object_init+0xb1/0x190 [nouveau]
Jun 09 15:31:02 dom0 kernel:  nouveau_do_resume+0x28/0x150 [nouveau]
Jun 09 15:31:02 dom0 kernel:  nouveau_pmops_runtime_resume+0x88/0x160 [nouveau]
Jun 09 15:31:02 dom0 kernel:  ? pci_restore_standard_config+0x40/0x40
Jun 09 15:31:02 dom0 kernel:  pci_pm_runtime_resume+0x74/0xd0
Jun 09 15:31:02 dom0 kernel:  __rpm_callback+0xc1/0x120
Jun 09 15:31:02 dom0 kernel:  ? pci_restore_standard_config+0x40/0x40
Jun 09 15:31:02 dom0 kernel:  rpm_callback+0x1f/0x70
Jun 09 15:31:02 dom0 kernel:  ? pci_restore_standard_config+0x40/0x40
Jun 09 15:31:02 dom0 kernel:  rpm_resume+0x513/0x690
Jun 09 15:31:02 dom0 kernel:  __pm_runtime_resume+0x47/0x70
Jun 09 15:31:02 dom0 kernel:  nouveau_drm_ioctl+0x35/0xc0 [nouveau]
Jun 09 15:31:02 dom0 kernel:  do_vfs_ioctl+0xab/0x690
Jun 09 15:31:02 dom0 kernel:  ? syscall_trace_enter+0x1c0/0x2d0
Jun 09 15:31:02 dom0 kernel:  ksys_ioctl+0x70/0x80
Jun 09 15:31:02 dom0 kernel:  __x64_sys_ioctl+0x16/0x20
Jun 09 15:31:02 dom0 kernel:  do_syscall_64+0x5b/0x170
Jun 09 15:31:02 dom0 kernel:  entry_SYSCALL_64_after_hwframe+0x44/0xa9
Jun 09 15:31:02 dom0 kernel: RIP: 0033:0x779a4fed76a7
Jun 09 15:31:02 dom0 kernel: Code: 00 00 00 48 8b 05 e1 27 2c 00 64 c7 00 26 00 00 00 48 c7 c0 ff ff ff ff c3 66 2e 0f 1f 84 00 00 00 00 00 b8 10 00 00 00 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 
Jun 09 15:31:02 dom0 kernel: RSP: 002b:00007fff2ea09d88 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
Jun 09 15:31:02 dom0 kernel: RAX: ffffffffffffffda RBX: 0000000000fb55d0 RCX: 0000779a4fed76a7
Jun 09 15:31:02 dom0 kernel: RDX: 00007fff2ea09dc0 RSI: 00000000c05064a7 RDI: 0000000000000011
Jun 09 15:31:02 dom0 kernel: RBP: 00007fff2ea09dc0 R08: 0000000000fb7da0 R09: 0000000000000000
Jun 09 15:31:02 dom0 kernel: R10: 0000000000000000 R11: 0000000000000246 R12: 00000000c05064a7
Jun 09 15:31:02 dom0 kernel: R13: 0000000000000011 R14: 00007fff2ea09dc0 R15: 0000000000000011
Jun 09 15:31:02 dom0 kernel: ---[ end trace c68ebc4d5296f569 ]---
Jun 09 15:31:02 dom0 kernel: nouveau 0000:01:00.0: tmr: stalled at ffffffffffffffff
```
This repeats until the log ends and the system crashes. The issue might also have been that I was running my Dom0 to the newest kernel version `Linux dom0 5.1.2-1.pvops.qubes.x86_64` while trying out some things a while ago. 
Once I disabled the nouveau drives I have had no problems. I dont know why the fix from #4891 seems to have stopped the issues for a while, maybe I was lucky.
